File: aoc/aoc2022day06tuningtrouble2.hs
-- uninhm
-- https://adventofcode.com/2022/day/6#part2
-- implementation, set

import qualified Data.Set as S

solve :: Int -> String -> Int
solve n s = go n (take n s) (drop n s)
  where
    go i lastN s | S.size (S.fromList lastN) == n = i
    go i (_:lastN') (x:s) = go (i+1) (lastN' ++ [x]) s

main :: IO ()
main = print . solve 14 =<< readFile "input"

File: aoc/aoc2022day06tuningtrouble2.cpp
// vicfred
// https://adventofcode.com/2022/day/6
// implementation
#include <fstream>
#include <iostream>
#include <set>

using namespace std;

int main() {
  ifstream input("input.txt");
  string line;
  getline(input, line);
  for (int i = 13; i < line.size(); ++i) {
    set<char> letters;
    for (int j = 0; j < 14; ++j) {
      letters.insert(line[i - j]);
    }
    if (letters.size() >= 14) {
      cout << i + 1 << endl;
      break;
    }
  }
  return 0;
}
